### PR TITLE
Validate keys in commands to make sure they do not exceed 250 characters

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR introduces a simple length check on keys that are provided as args to various memcached commands.

### Details - How are you making this change?  What are the effects of this change?

In the case of a write command (`set`, `add` etc.) being issued with a key that[ exceeds memcached's limit of 250 characters](https://github.com/memcached/memcached/blob/5609673ed29db98a377749fab469fe80777de8fd/doc/protocol.txt#L44), the server will behave in a way that interrupts expected operation of this crate (and others).  Example of this behaviour:

```
$ memcached -vvv

# in another terminal:
$ echo "set keyaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 0 0 5\r\nvalue\r\n" | nc localhost 11211

# output from memcached server in vvv mode:

<27 new auto-negotiating client connection
27: going from conn_new_cmd to conn_waiting
27: going from conn_waiting to conn_read
27: going from conn_read to conn_parse_cmd
27: Client using the ascii protocol
<27 set keyaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 0 0 5
>27 CLIENT_ERROR bad command line format
27: going from conn_parse_cmd to conn_new_cmd
27: going from conn_new_cmd to conn_parse_cmd
<27 value
>27 ERROR
27: going from conn_parse_cmd to conn_new_cmd
27: going from conn_new_cmd to conn_parse_cmd
<27
>27 ERROR
27: going from conn_parse_cmd to conn_new_cmd
27: going from conn_new_cmd to conn_mwrite
27: going from conn_mwrite to conn_new_cmd
27: going from conn_new_cmd to conn_waiting
27: going from conn_waiting to conn_read
27: going from conn_read to conn_closing
<27 connection closed.
27: going from conn_closing to conn_closed
```
From this output we can see that the memcached server parses a `set` command in a few steps:
1) Reads the incoming command from a buffer until it reaches the first `\r\n` delimiter, i.e. reads up to `set keyaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 0 0 5`
2) Determines that the key is too long and returns `CLIENT_ERROR bad command line format`
3) Continues to parse what's left on the buffer until the second `\r\n` delimiter, i.e. `value`
4) Interprets what it has read as another incoming command rather than the remnants of a bad `set`, which prompts it to return `ERROR` to the client, as a bad/malformed command.

This leaves an unexpected error on the buffer which is not properly read.  That error remaining on the buffer will instead be incorrectly parsed by the next successful command that is issued, leading to unexpected behaviour and commands being interpreted as failed inappropriately.  This must either be handled by the crate or the crate consumer, and since it is a problem that affects all users, it should be handled on the crate level.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -170,7 +170,9 @@ impl Client {
     {
         self.conn.write_all(b"get ").await?;
         for key in keys {
-            self.conn.write_all(key.as_ref()).await?;
+            self.conn
+                .write_all(Self::validate_key_length(key.as_ref())?)
+                .await?;
```
[the remark this answers is not in the document]
Comment (grcooper) [reply src/lib.rs:175]: Yah I like that idea, but we only return a single error right, we can't map that error to our key currently?

File: src/lib.rs [hunk 2]
```diff
@@ -168,10 +168,14 @@ impl Client {
...
-        self.conn.write_all(b"get ").await?;
+        self.conn.write_all(b"get").await?;
         for key in keys {
-            self.conn.write_all(key.as_ref()).await?;
+            const MAX_KEY_LENGTH: usize = 250;
+            if key.as_ref().len() > MAX_KEY_LENGTH {
+                continue;
+            }
             self.conn.write_all(b" ").await?;
+            self.conn.write_all(key.as_ref()).await?;
```
Comment (mrattle) [inline src/lib.rs:178]: Fixed this to remove the trailing whitespace at the end of the command to compensate for [this rollback](https://github.com/Shopify/async-memcached/pull/61).

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (grcooper) [reply src/lib.rs]: Yah I like the idea of doing best-attempt in a set_multi 👍 

For the non-multi ones I would return an error though.

[the remark this answers is not in the document]
Comment (mrattle) [reply src/lib.rs]: Actually, on that point we could maybe populate our own error in the response hashmap for a `multi_set`, and just do best effor for `multi_get`.

File: src/lib.rs [hunk 3]
```diff
@@ -20,6 +20,8 @@ pub use self::parser::{ErrorKind, KeyMetadata, MetadumpResponse, StatsResponse,
 mod value_serializer;
 pub use self::value_serializer::AsMemcachedValue;
 
+const MAX_KEY_LENGTH: usize = 250;
```
Comment (nickamorim) [inline src/lib.rs:23]: Might be worth linking to the Memcached documentation where this is defined

File: src/lib.rs [hunk 4]
```diff
@@ -109,6 +111,17 @@ impl Client {
         let mut results = FxHashMap::with_capacity_and_hasher(kv.len(), Default::default());
 
         for (key, _) in kv {
+            let kr = key.as_ref();
+            if kr.len() > MAX_KEY_LENGTH {
+                results.insert(
+                    key,
+                    Err(Error::Protocol(Status::Error(ErrorKind::Client(
+                        "Key exceeds maximum length of 250 bytes".to_string(),
```
Comment (nickamorim) [inline src/lib.rs:119]: Should we define a custom error kind instead?

**Review by nickamorim — COMMENTED** [review]
I'm personally not a fan of a best-effort solution. I think we should just fail the entire request if a key exceeds the max length. If we just silent swallow these errors and treat as a cache miss / don't try to store, it's very confusing to users as to why their key wasn't stored or missed. Also, typically people would log / trace errors so they can fix them going forward (i.e. by hashing the key). If we just swallow the errors, keys could be silently failing forever without any user feedback

File: src/lib.rs (hunk 3 above)
Comment (mrattle) [reply src/lib.rs]: Done.

File: src/lib.rs (hunk 4 above)
Comment (mrattle) [reply src/lib.rs:119]: Updated with a custom `KeyTooLong` entry in `ErrorKind` which replaces the `Client` error.  Name sound fine?

**Review by danmayer — APPROVED** [review]
